content/posts/2025-gsoc-clang-doc.md

```diff
@@ -0,0 +1,375 @@
+At one time, YAML was meant to be that format, but it suffered from feature disparity since it wasn't relevant when something needed to be displayed in HTML.
+If we used HTML instead, there was a lot of other data (tags, indentation, classes, IDs) that would need to be validated alongside the construct.
+Testing the documentation and testing the displayed content are two different tasks.
+I ended up adding 14 different test files over the course of the summer to ensure test coverage.
+
+### Pull Requests
+- [add tags to Mustache namespace template](https://github.com/llvm/llvm-project/pull/142045)
+- [add a JSON generator](https://github.com/llvm/llvm-project/pull/142483)
+- [add namespaces to JSON generator](https://github.com/llvm/llvm-project/pull/143209)
+- [removed default label on some switches](https://github.com/llvm/llvm-project/pull/143919)
+- [precommit](https://github.com/llvm/llvm-project/pull/144160) and [add support for concepts](https://github.com/llvm/llvm-project/pull/144430)
+- [precommit](https://github.com/llvm/llvm-project/pull/145069) and [document global variables](https://github.com/llvm/llvm-project/pull/145070)
+- [serialize isBuiltIn and IsTemplate](https://github.com/llvm/llvm-project/pull/146149)
+- [precommit](https://github.com/llvm/llvm-project/pull/146164) and [serialize friends](https://github.com/llvm/llvm-project/pull/146165)
+
+# Comments
+
+## Groups and Order
+
+Comments weren't ordered in Clang-Doc's HTML documentation.
+They were just displayed in whatever order they were serialized in, which is the order that they're written in source.
+This meant comments would be extremely difficult to read - you don't want to search for another parameter comment after reading the first one, even if they're expected to be written in order in source.
+
+Funnily enough, Mustache made this a little more complicated.
+The only logic operation that Mustache has to check if a field exists is an iteration like `{{#Fields}}`, but any header that denotes a comment section would be duplicated.
+
+```html
+{{#Fields}}
+<h3>Field Header</h3>
+  {{FieldInfo}}
+{{/Fields}}
+```
+
+the `<h3>` header would be duplicated for every iteration over `Fields`.
+If the header was outside of the iteration, then it would be displayed even if there weren't any elements in `Fields`.
+All of the logic to order them needs to be done in the serialization to JSON itself, so I had overhaul our comment organization.
+
+Previously, Clang-Doc's comments were organized exactly as in Clang's AST like the following:
+
+- FullComment
+  - BriefComment
+    - ParagraphComment
+      - TextComment
+      - TextComment
+  - BriefComment
+    - ParagraphComment
+
+Everything was unnecessarily nested under a FullComment, and TextComments were also unnecessarily nested.
+Every non-verbatim comment's text was held in one ParagraphComment.
+Since there was only one, we could reduce some boilerplate by directly mapping to the array of TextComments.
```

Comment:
```suggestion
Everything was unnecessarily nested under a `FullComment`, and `TextComment`s were also unnecessarily nested.
Every non-verbatim comment's text was held in one `ParagraphComment`.
Since there was only one, we could reduce some boilerplate by directly mapping to the array of `TextComment`s.
```